File: schemas/authentication.py
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, Field

from settings.settings import TimeZoneSettings

time_zone_settings = TimeZoneSettings()

def get_current_time():
    return datetime.now(time_zone_settings.ZURICH_TZ)

# Won't be sent
class TokenDataSchemaBase(BaseModel):
    user_id: int
    email: str
    roles: List[str]  # Expecting a list of roles

    model_config = {
        "from_attributes": True
    }

class TokenDataSchemaFull(TokenDataSchemaBase):

    model_config = {
        "from_attributes": True
    }

# Send user details
class UserDataSchemaBase(BaseModel):
    user_id: int
    email: str
    roles: List[str]  # Expecting a list of roles

    model_config = {
        "from_attributes": True
    }

class UserDataSchemaFull(UserDataSchemaBase):
    login_at: Optional[datetime] = Field(default_factory=get_current_time)

    model_config = {
        "from_attributes": True
    }

# Send back access token for user
class LoginSchemaBase(BaseModel):
    access_token: str
    token_type: str

    model_config = {
        "from_attributes": True
    }

class LoginSchemaFull(LoginSchemaBase):
    login_at: Optional[datetime] = Field(default_factory=get_current_time)

    model_config = {
        "from_attributes": True
    }

File: frontend/types/authentication.ts
export interface TokenDataSchemaBase {
  user_id: number;
  email: string;
  roles: string[]; // List of roles
}

export interface TokenDataSchemaFull extends TokenDataSchemaBase {
  // No additional fields in the derived schema
}

export interface UserDataSchemaBase {
  user_id: number;
  email: string;
  roles: string[]; // List of roles
}

export interface UserDataSchemaFull extends UserDataSchemaBase {
  login_at?: Date; // Optional datetime field
}

export interface LoginSchemaBase {
  access_token: string;
  token_type: string;
}

export interface LoginSchemaFull extends LoginSchemaBase {
  login_at?: Date; // Optional datetime field
}
